### DISCUSSION TITLE
How to read AggregateFunction(groupBitmap, UInt32) column directly

### DISCUSSION THREAD

**Discussion by xichenAiguang** [DarkWanderer/ClickHouse.Client#513] [Q&A] [answered]
Hi all, I have a job ingest data into Clickhouse, which has a RoaringBitmap field, its type is AggregateFunction(groupBitmap, UInt32) in Clickhouse. I encoutered this exception " _Unknown type: AggregateFunction(groupBitmap, UInt32)_" when calling ExecuteReader(). I understand it's due to lack of support of this type, I'd like to extend TypeConverter to support it, but seems like it's impossible because TypeConverter  is a internal class. My questions is how to read this column, note that I don't want to convert this column to String or Bit or other datatypes with Clickhouse built-in function like ToString(), bitmapToArray, etc. Thanks in advance for any suggestions

**Comment by DarkWanderer** [comment]
Hi. As with all `AggregateFunction` fields, when selecting the data, you need to use appropriate `*Merge` function - see [docs](https://clickhouse.com/docs/en/sql-reference/data-types/aggregatefunction). For `groupBitmap`, the correct function would be `groupBitmapMerge`: 
```
SELECT groupBitmapMerge(myColumn) FROM myTable
```

**Reply by xichenAiguang** [reply]
Thank you for prompt reply.  The function groupBitmapMerge returns the cardinality of that Bitmap, not Bitmap itself. I'd like to query the Bitmap from my  service, and deserialize to C# RoaringBitmap instance. To achieve that, we need to call groupBitmapMergeState instead, unfortunately  hit  the exception " Unknown type: AggregateFunction(groupBitmap, UInt32)" again.

**Reply by xichenAiguang** [reply]
Attach the image for easier understanding. My purpose is that get ClickhouseRoaringBitmap value

![image](https://github.com/user-attachments/assets/62a31968-8b33-40db-bb0f-16e5408d51c4)

**Reply by DarkWanderer** [reply, marked as answer]
I don't yet understand the purpose of selecting raw AggregateFunction state, but if you would really like to read it, there is a function [ExecuteRawResultAsync](https://github.com/DarkWanderer/ClickHouse.Client/blob/7c953726b3da9fc271df772191e6d3064ae5d396/ClickHouse.Client/ADO/ClickHouseCommand.cs#L98) which allows to read raw column values directly - you just need to specify a `FORMAT` you can parse

ClickHouse support channel may be a good help in finding a better way - https://telegram.me/clickhouse_en

**Reply by xichenAiguang** [reply]
Got it, thank you for pointing out. Wondering do you have any plan support this type?

**Reply by DarkWanderer** [reply]
No, as the internals of aggregation function state are not supposed to be used directly, and are not part of ClickHouse interface I can use

**Comment by DarkWanderer** [comment]
Since this seems to be a common question, I'll add a helper function which gives a more appropriate exception text in next release
